### PR TITLE
✨Amp-consent CCPA geoOverride and consentRequired

### PR DESCRIPTION
**Pull request by micajuine-ho**
OUTDATED:
2nd and 4th task in #25623. Follow up to #25676

- Support `geoOverride` and `consentRequired` in config.
- Added `matchedGeoGroups` field to `checkConsentHref` request.

### CODE REVIEW

File: extensions/amp-consent/0.1/amp-consent.js [hunk 1]
```diff
@@ -196,10 +218,47 @@ export class AmpConsent extends AMP.BaseElement {
       this.notificationUiManager_ = /** @type {!NotificationUiManager} */ (manager);
     });
 
+    // (TODO): Assert that if we're using new version we have consentRequired
+    // otherwise we're using old version and we should migrate to new
+    // version here.
+    // E.g: promptIfUnknownForGeo: geoGroup -> geoOverride: {geoGroup: {consentRequired: true}}
+
+    // Iterate through geoGroups, and override toplevel `checkConsentHref` and `consentRequire` if we are in the group
+    const geoServicePromise = this.consentConfig_['geoOverride']
```
Comment (lannka) [inline extensions/amp-consent/0.1/amp-consent.js]: this logic can be moved to `consent-config.js`. it might deserve its own PR.

File: extensions/amp-consent/0.1/amp-consent.js [hunk 2]
```diff
@@ -521,7 +506,9 @@ export class AmpConsent extends AMP.BaseElement {
           'consentInstanceId': this.consentId_,
           'consentStateValue': getConsentStateValue(storedInfo['consentState']),
           'consentString': storedInfo['consentString'],
+          // isDirty will be deprecated
```
Comment (lannka) [inline extensions/amp-consent/0.1/amp-consent.js]: I'm actually not sure about this. Need @zhouyx 's input later. But let's not assume this is going away for now.

File: extensions/amp-consent/0.1/amp-consent.js [hunk 3]
```diff
@@ -547,40 +534,98 @@ export class AmpConsent extends AMP.BaseElement {
...
+  /**
+   * Sync client side storage with server response, if applicable.
+   *
+   * @param {?Object} response
+   */
+  updateCache(response) {
+    // If we do prompt then we should NOT sync with server (except to clear cache).
+    if (response) {
+      // Need to set dirty bit first, so that updated state will also include it
+      if (!!response['expireCache']) {
```
Comment (lannka) [inline extensions/amp-consent/0.1/amp-consent.js]: let's be explicit:

```
response['expireCache'] === true
```

File: extensions/amp-consent/0.1/amp-consent.js [hunk 4]
```diff
@@ -547,40 +534,98 @@ export class AmpConsent extends AMP.BaseElement {
...
+    }
+  }
+
+  /**
+   * Attempts to sync with the server endpoint if there is one.
+   */
+  syncServerData_() {
+    // This is where passSharedData and clearCache will be as well as syncing with the cache
+    this.getConsentRemote_().then(response => {
+      // Pass shared data
```
Comment (lannka) [inline extensions/amp-consent/0.1/amp-consent.js]: no need to over comment on the code. the method names are already explaining.
this applies to all the code

File: extensions/amp-consent/0.1/amp-consent.js [hunk 5]
```diff
@@ -547,40 +534,98 @@ export class AmpConsent extends AMP.BaseElement {
...
-    return this.consentStateManager_.getConsentInstanceInfo().then(info => {
-      if (hasStoredValue(info)) {
-        // Has user stored value, no need to prompt
-        return true;
+  /**
+   * Blindly pass sharedData
+   *
+   * @param {?Object} response
+   */
+  passSharedData_(response) {
```
Comment (lannka) [inline extensions/amp-consent/0.1/amp-consent.js]: try not to move methods around in this PR. it messed up the code diff

**Comment by micajuine-ho** [comment]
Currently splitting this PR into smaller sections...
